rfd/XXXXa-scoped-join-tokens.md

```diff
@@ -0,0 +1,236 @@
++
++  // The number of resources that can be provisioned using this token.
++  int32 max_uses = 6;
++
++  // The number of resources that have been provisioned using this token.
++  int32 total_uses = 6;
+ }
+```
+
+### Provisioning
+
+Scoped resources will be provisioned by extending the existing `Join` RPC to
+support scoped tokens. When the join method is `token` the auth server will
+query both the existing `ProvisionTokenV2` and the new `ScopedToken` resources
+using the provided token name. If a `ScopedToken` is found then it will take
+precedence over any `ProvisionTokenV2` that might share the same name. This
+approach allows for backwards compatible provisioning where agents can be
+scoped without any knowledge of scopes themselves. The `assigned_scope` of the
+`ScopedToken` will be assigned as the scope of the resulting resource. It will
+also be attached as metadata to the resulting host certificate under the
+`AgentScope` field so that scope related access controls can be performed
+against the agent's identity rather than just restricting clients' access to
+the agent itself.
+
+### Automatic Labels
+
+Similar to assigning the provisioned resource scope from the `ScopedToken`
+used, the same token's `assigned_labels` field will be applied to the resulting
+resource directly after provisioning. Because static labels exist exclusively
+as configuration within the agent's `teleport.yaml` file, scoped token
+provisioning will use the `server_info` resource to apply
+[resource-based labels](https://goteleport.com/docs/zero-trust-access/rbac-get-started/labels/#apply-resource-based-labels)
+upon creation. This applies labels at runtime using the inventory control
+stream and does not require any new resources or RPCs. Labels will be created
+at the end of [Join()](https://) before returning the host certificate for a
+successfully provisioned resource.
+
+### Limited Use Tokens
+
+The simplest way to implement token usage limits is to keep a counter of
+successful uses and fallback on the conditional update system to maintain
+consistency. Whenever a resource is provisioned using a token, we increment the
+token's `total_uses` field with a conditional update. If the `total_uses`
+reaches the `max_uses`, the token will no longer be usable and should
+eventually be cleaned up. This is simple to reason about but requires that the
+final decision about whether or not the token can be used has to be deferred to
+the end of the provisioning process. Otherwise it would be possible for many
+concurrent join attempts to exceed the allowed `max_uses` for a token.
+
+To reduce the throughput to the backend, each auth server would queue attempts
+to update the token and flush them as a single update on a set interval. The
+interval would need to be a relatively short as it would increase provisioning
+time by the interval itself in the worst case. This way writes could be batched
+per auth server per interval instead of per provisioning attempt.
```

Comment:
nit: this is an interesting idea, but I'd suggest not including it in the initial implementation.  We can always circle back to it, and my gut feeling is that, where high volume joins are concerned, we probably want to continue to recommend unlimited use tokens.

Reply:
Please at least add a mutex to the initial implementation so the same auth can't interfere with itself while updating the token status.